common/rsyslog_plugin.cpp

```diff
@@ -0,0 +1,161 @@
+#include <iostream>
+#include <vector>
+#include <string>
+#include <cstdlib>
+#include <cstring>
+#include <fstream>
+#include <unordered_set>
+#include <regex>
+#include <tuple>
+#include "logger.h"
+#include "json.hpp"
+
+using namespace std;
+using namespace nlohmann;
+
+string parserCfgFile = "{}_parse.rc.json";
+ofstream outfile;
+string logfile = "/tmp/{}_logfile";
+json regexList = json::array();
+string pgmName = "";
+unordered_set<string> errReportedTags;
+
+void loadRegexlist(string rcPath, string pgmName) {
+	fstream rcfile;
+	parserCfgFile = regex_replace(parserCfgFile, regex("\\{}"), pgmName);
+	parserCfgFile = rcPath + parserCfgFile;
+	rcfile.open(parserCfgFile, ios::in);
+	if (!rcfile) {
+		cerr << "No such file: " << parserCfgFile << endl;
+		return;
+	}
+	rcfile >> regexList;
+}
+
+tuple<bool, json> matchRegex(string msg) {
+	bool isMatch = false;
+	json structuredEvent;
+	for (int i = 0; i < regexList.size(); i++) {
+		string regexString = regexList[i]["regex"];
+		regex expression(regexString);
+		match_results<string::const_iterator> matchResults;
+		regex_search(msg, matchResults, expression);
+		vector<string> groups;
+		// first match in groups is entire msg string
+		for(int j = 1; j < matchResults.size(); j++) {
+			groups.push_back(matchResults.str(i));
+		}
+		vector<string> params = regexList[i]["params"];
+		string tag = regexList[i]["tag"];
+		structuredEvent["tag"] = tag;
+		structuredEvent["program"] = pgmName;
+		if(params.size() != groups.size() && errReportedTags.find(tag) == errReportedTags.end()) {
+			errReportedTags.insert(tag);
+			structuredEvent["parsed"] = groups;
+		} else {
+			for(int j = 0; j < params.size(); j++) {
+				structuredEvent[params[j]] = "groups[j]";
+			}
+			isMatch = true;
+			break;
+		}
+	}
+	return make_tuple(isMatch, structuredEvent);
+}
+
+void onInit(bool isDebug, string rcPath, string pgmName) {
+	loadRegexlist(rcPath, pgmName);
+	if (isDebug) {
+		logfile = regex_replace(logfile, regex("\\{}"), pgmName);
+		outfile.open(logfile);
+	}
+}
+
+void onMessage(string msg) {
+	tuple<bool, json> regexMatch = matchRegex(msg);
+	bool match = get<0>(regexMatch);
+	json event = get<1>(regexMatch);
+
+	if(outfile) {
+		outfile << "match: " << match << "\noriginal message: " << msg << "\nstructured event: " << event << endl;
+	}
+
+	if(match) {
+		// call event api with event
+	}
+}
+
+void onExit() {
+	if(outfile) {
+		outfile.close();
+	}
+}
+
+void run() {
+	while(true) {
+		string line;
+		getline(cin, line);
+		if(line.empty()) {
+			continue;
+		}
+		onMessage(line);
+		cout << "Called onMessage on: " << line << endl;
+	}
+}
+
+void showUsage(string name) {
+	cerr << "Usage: " << name << " <option(s)> SOURCES\n"
+		<< "Options:\n"
+		<< "\t-h,--help\t\tShow this help message\n"
+		<< "\t-d,--debug,type=bool\t\tRun with debug log level"
+		<< "\t-p,--pgm-name,type=string,required=True\t\tName of the program dumping the message"
```

Comment:
>Name of the program dumping the message [](http://example.com/codeflow?start=51&length=39)

Why ask user about this?

Reply:
The user in this case will be rsyslog configuration file. Depending on the process that we want to send syslogs for such as bgpd or bgpcfgd, we will create a new plugin instance for that specific program.

You can take a look at example configuration file:

https://github.com/zbud-msft/sonic-buildimage/pull/1/files#diff-6c6cef737a73f930b17f7bb06fde12832fc17b5c8a17d5422cbebb5d9e387757

Reply:
if `-r` is full path, you do not need a standalone `-p`.

Reply:
We need the source name for the event API call that we will be making, as well as if we want to write to a logfile. 

We can either get it from -r path and parse the source name, or we can parse from syslog. For each syslog message, are we guaranteed to get source name? I think it would be better to get once from syslog message and store it, versus dealing with parsing from rcpath.